components/movementsensor/gpsutils/cachedData_test.go

```diff
@@ -60,3 +63,171 @@ func TestReadingsSerial(t *testing.T) {
 	test.That(t, err, test.ShouldBeNil)
 	test.That(t, fix1, test.ShouldEqual, fix)
 }
+
+func TestPosition(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+	g := NewCachedData(&mockDataReader{}, logger)
+
+	t.Run("no current location", func(t *testing.T) {
+		g.lastPosition.SetLastPosition(geo.NewPoint(32.4, 54.2))
+		g.nmeaData = NmeaParser{
+			Location: nil,
+		}
+
+		expectedPoint := geo.NewPoint(32.4, 54.2)
+
+		pos, alt, err := g.Position(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, movementsensor.ArePointsEqual(pos, expectedPoint), test.ShouldBeTrue)
+		test.That(t, alt, test.ShouldEqual, 0.0)
+	})
+
+	t.Run("current location zero so return last known", func(t *testing.T) {
+		g.lastPosition.SetLastPosition(geo.NewPoint(32.4, 54.2))
+		g.nmeaData = NmeaParser{
+			Location: geo.NewPoint(0, 0),
+			Alt:      12.1,
+		}
+
+		expectedPoint := geo.NewPoint(32.4, 54.2)
+
+		pos, alt, err := g.Position(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, movementsensor.ArePointsEqual(pos, expectedPoint), test.ShouldBeTrue)
+		test.That(t, alt, test.ShouldEqual, 12.1)
+
+		// Check that the last known position was not updated
+		test.That(t, movementsensor.ArePointsEqual(g.lastPosition.GetLastPosition(), expectedPoint), test.ShouldBeTrue)
+	})
+
+	t.Run("Valid current location", func(t *testing.T) {
+		g.lastPosition.SetLastPosition(geo.NewPoint(32.4, 54.2))
+		g.nmeaData = NmeaParser{
+			Location: geo.NewPoint(1.1, 1.2),
+			Alt:      1.3,
+		}
+
+		expectedPoint := geo.NewPoint(1.1, 1.2)
+
+		pos, alt, err := g.Position(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, movementsensor.ArePointsEqual(pos, expectedPoint), test.ShouldBeTrue)
+		test.That(t, alt, test.ShouldEqual, 1.3)
+
+		// Check that the last known position was updated
+		test.That(t, movementsensor.ArePointsEqual(g.lastPosition.GetLastPosition(), expectedPoint), test.ShouldBeTrue)
+	})
+}
+
+func TestLinearVelocity(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+	g := NewCachedData(&mockDataReader{}, logger)
+
+	t.Run("no compass heading", func(t *testing.T) {
+		g.nmeaData = NmeaParser{
+			CompassHeading: math.NaN(),
+		}
+
+		speed, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, speed.X, test.ShouldEqual, 0.0)
+		test.That(t, speed.Y, test.ShouldEqual, 0.0)
+	})
+
+	t.Run("test sample velocity", func(t *testing.T) {
+		// Two quandrants of the compass
+		// Enough to verify quadrant signs are correct
```

Comment:
Elaborate more on this comment here. Which two quadrants and why is it enough to verify quadrant signs and what happens if we add values from other quads?

Reply:
Yeah, since you just need to verify {(+,+) or (-,-)} and {(+,-) or (-,+)}, which allows you to get the relevant quadrants + the orientation (clockwise, counterclockwise). I can make the comment more detailed, or I can add tests for the other quadrants @susmitaSanyal

Reply:
i'll just add other quadrants for robustness

Reply:
do both. the comment should explain "why" we are testing faulty values.